tfimm/architectures/lora/factory.py

```diff
@@ -149,41 +150,97 @@ def convert_to_regular_model(model: tf.keras.Model) -> tf.keras.Model:
-    for layer in model._flatten_layers(recursive=True, include_self=False):
+    trainable_weights = []
+    for layer in model._flatten_layers(recursive=True, include_self=True):
         if getattr(layer, "is_lora_layer", False):
-            layer.set_only_lora_weights_trainable(train_bias in {"all", "lora_only"})
+            trainable_weights.extend(
+                layer.lora_trainable_weights(train_bias in {"all", "lora_only"})
+            )
         elif train_bias in {"all"}:
-            _set_bias_weights_trainable(layer)
+            trainable_weights.extend(_bias_variables(layer))
+
+    return trainable_weights
 
 
-def _set_bias_weights_trainable(layer: tf.keras.layers.Layer):
-    # TODO: Implement setting biases only to be trainable.
-    raise NotImplementedError("Need to implement this...")
+def lora_non_trainable_weights(
+    model: tf.keras.Model, train_bias: str = "none"
+) -> List[tf.Variable]:
+    """
+    Returns a list of non-trainable weights for the LoRA model. This function
+    complements :py:func:`lora_trainable_weights`.
+
+    Args:
+        model: A keras model.
+        train_bias: If "none" or "all", no or all bias weights are trainable
+            respectively. If "lora_only", only the bias weights of LoRA layers are set
+            to trainable.
+
+    Returns:
+        List of LoRA non-trainable weights.
+    """
+    # We start with a list of all weights and remove the trainable weights. We
+    # explicitly call `(non_)trainable_weights` from tf.keras.Model, because we expect
+    # our model to overwrite the `(non_)trainable_weights` properties with the result
+    # of this function.
+    weights = tf.keras.Model.trainable_weights.fget(
+        model
+    ) + tf.keras.Model.non_trainable_weights.fget(model)
+    trainable_weights = lora_trainable_weights(model, train_bias=train_bias)
+    trainable_ids = set(id(w) for w in trainable_weights)  # Variables are not hashable
+    non_trainable_weights = [w for w in weights if id(w) not in trainable_ids]
+    return non_trainable_weights
+
+
+def _bias_variables(layer: tf.keras.layers.Layer) -> List[tf.Variable]:
+    """Return bias variables of a given layer as a list."""
+    # There is no fundamental concept of a "bias-variable", rather the notion of bias
+    # is a convention. And so we have to proceed layer-by-layer and specify for each
+    # layer, which are the bias variables (and if they are used).
+    if (
+        type(layer)
+        in {
+            tf.keras.layers.Conv1D,
+            tf.keras.layers.Conv2D,
+            tf.keras.layers.Conv3D,
+            tf.keras.layers.Dense,
+            tf.keras.layers.DepthwiseConv1D,
+            tf.keras.layers.DepthwiseConv2D,
+        }
+        and layer.use_bias
+    ):
+        return [layer.bias]
+    elif (
+        type(layer)
+        in {
+            tf.keras.layers.BatchNormalization,
+            tf.keras.layers.LayerNormalization,
```

Comment:
does it make sense to freeze normalisation layers? if frozen, they will not really "normalise" their inputs?

Reply:
Normalisation layers contain two parts: First, they normalise their inputs. But then, they contain learnable scale and bias parameters to map the normalised inputs to the desired mean and bias. So here we are saying that during the training we continue to train the bias parameter. Normalization happens regardless.